### PR TITLE
Full modem update support for lwm2m_client

### PR DESCRIPTION
**Pull request by jarlamsa**
Add support for using the full modem update on the lwm2m_client.
Currently it needs an external flash to store the firmware update image.

Both the out-of-band and in-band update mechanisms are supported.
After downloading the firmware update to the flash, the update resource needs to be executed.
The update is applied after the execute is called and then the device reboots with the new modem firmware.

### CODE REVIEW

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 1]
```diff
@@ -445,6 +447,13 @@ void main(void)
 		return;
 	}
 
+#if !defined(CONFIG_NRF_MODEM_LIB_SYS_INIT)
+	ret = nrf_modem_lib_init(NORMAL_MODE);
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply samples/nrf9160/lwm2m_client/src/main.c:451]: `FMFU_DEV` needs the configuration option `!NRF_MODEM_LIB_SYS_INIT`, so therefore it is needed to use the manual initialization to use the full modem update.

File: samples/nrf9160/lwm2m_client/boards/nrf9160dk_nrf9160_ns.conf [hunk 2]
```diff
@@ -18,3 +18,36 @@ CONFIG_LWM2M_IPSO_ONOFF_SWITCH_VERSION_1_1=y
...
+
+# Full modem update
+CONFIG_DFU_TARGET_FULL_MODEM=y
+CONFIG_FMFU_FDEV=y
+CONFIG_CDDL_GEN=y
+CONFIG_MODEM_INFO=y
+CONFIG_FMFU_FDEV_SKIP_PREVALIDATION=n
+
+# Crypto
+CONFIG_OBERON_BACKEND=y
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply samples/nrf9160/lwm2m_client/boards/nrf9160dk_nrf9160_ns.conf:40]: Might be that isn't needed, the crypto-options came from the full_modem_update -sample. Need to test the config options.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 3]
```diff
@@ -76,23 +151,16 @@ static int firmware_update_cb(uint16_t obj_inst_id, uint8_t *args,
 		goto cleanup;
 	}
 
-	if (!fota_download_finished) {
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:79]: True, deleted

**Comment by jarlamsa** [comment]
Thank you for your reviews @rlubos and @umapraseeda I think all the comments have been addressed. Please re-review.

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

File: doc/nrf/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -43,6 +43,7 @@ nRF9160
     * Added more LwM2M objects.
     * LwM2M sensor objects now uses the actual sensors available to the Thingy:91. If the nRF9160 DK is used, it uses simulated sensors instead.
     * Added possibility to poll sensors and notify the server if the measured changes are large enough.
+    * Added possibility to use full modem firmware update.
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:46]:
```suggestion
    * Added support for polling sensors and notifying the server if the measured changes are large enough.
    * Added support for full modem firmware update.
```

Comment (jarlamsa) [reply doc/nrf/releases/release-notes-changelog.rst]: The upper line wasn't part of my PR, but changed that also to the suggested format.

**Pull request merged**